tools/codegen/core/gen_huffman_decompressor.cc

```diff
@@ -1625,23 +1640,38 @@ std::string SplitBefore(absl::string_view input, char c) {
 
 // Does what it says.
 void WriteFile(std::string filename, std::string content) {
+  auto out = grpc_core::GetEnv("GEN_OUT");
+  if (out.has_value()) {
+    filename = absl::StrCat(*out, "/", filename);
+  }
   std::ofstream ofs(filename);
   ofs << content;
+  if (ofs.bad()) {
+    fprintf(stderr, "Failed to write %s\n", filename.c_str());
+    abort();
+  }
 }
 
-int main(void) {
-  std::vector<std::unique_ptr<BuildOutput>> results;
+void GenMicrobenchmarks() {
   std::queue<std::thread> threads;
   // Generate all permutations of max_bits_for_depth for the Build function.
   // Then generate all variations of the code.
+  static constexpr int kNumShards = 100;
+  std::unique_ptr<FileSet> results[kNumShards];
+  std::mutex results_mutexes[kNumShards];
+  for (int i = 0; i < kNumShards; i++) {
+    results[i] = std::make_unique<FileSet>(
+        absl::StrCat("test/cpp/microbenchmarks/huffman_geometries/shard_", i));
+    results[i]->AddFrontMatter(2024);
+  }
+  int r = 0;
   for (const auto& perm : PermutationBuilder(3).Run()) {
-    while (threads.size() > 200) {
-      threads.front().join();
-      threads.pop();
-    }
-    results.emplace_back(std::make_unique<BuildOutput>());
+    int shard = r++ % kNumShards;
     threads.emplace(
-        [perm, r = results.back().get()] { *r = Build(perm, false, 2023); });
+        [perm, fileset = results[shard].get(), mu = &results_mutexes[shard]] {
+          std::lock_guard<std::mutex> lock(*mu);
```

Comment:
Do `std::mutex` and `std::lock_guard` this work everywhere we need to build/test?

Reply:
everywhere = linux, and yes (trying to keep this code low dependency for the moment)